addons/iap/models/iap_account.py

```diff
@@ -17,103 +18,121 @@
+                    'warning_threshold': account.warning_threshold,
+                    'warning_emails': [{
+                        'email': user.email,
+                        'lang_code': user.lang or get_lang(self.env).code,
+                    } for user in account.warning_user_ids],
                 }
-                iap_tools.iap_jsonrpc(url=url, params=data)
-            except AccessError as e:
-                _logger.warning('Save service error : %s', str(e))
+                try:
+                    iap_tools.iap_jsonrpc(url=url, params=data)
+                except AccessError as e:
+                    _logger.warning("Update of the warning email configuration has failed: %s", str(e))
         return res
 
-    def get_services(self):
+    @staticmethod
+    def is_running_test_suite():
+        return hasattr(threading.current_thread(), 'testing') and threading.current_thread().testing
+
+    def _get_account_information_from_iap(self):
+        # During testing, we don't want to call the iap server
+        if self.is_running_test_suite():
+            return
+        route = '/iap/1/get-accounts-information'
+        endpoint = iap_tools.iap_get_endpoint(self.env)
+        url = werkzeug.urls.url_join(endpoint, route)
+        params = {
+            'iap_accounts': [{
+                'token': account.account_token,
+                'service': account.service_id.technical_name,
+            } for account in self],
+            'dbuuid': self.env['ir.config_parameter'].sudo().get_param('database.uuid'),
+        }
         try:
-            route = '/iap/services-token'
-            endpoint = iap_tools.iap_get_endpoint(self.env)
-            url = endpoint + route
-            account_tokens = self.env['iap.account'].sudo().search([]).mapped('account_token')
-            params = {
-                'dbuuid': self.env['ir.config_parameter'].sudo().get_param('database.uuid'),
-                'iap_accounts': account_tokens,
-            }
-            services = iap_tools.iap_jsonrpc(url=url, params=params)
-            for service in services:
-                account_id = self.env['iap.account'].sudo().search(
-                    [('account_token', '=', service['account_token'])]).ids[0]
-                service['account_id'] = account_id
-                self.env['iap.account.info'].create(service)
+            accounts_information = iap_tools.iap_jsonrpc(url=url, params=params)
         except AccessError as e:
-            _logger.warning('Get services error : %s', str(e))
+            _logger.warning("Fetch of the IAP accounts information has failed: %s", str(e))
+            return
+
+        for token, information in accounts_information.items():
+            account_id = self.filtered(lambda acc: secrets.compare_digest(acc.account_token, token))
+
+            balance_amount = round(information['balance']) if account_id.service_id.integer_balance else information['balance']
```

Comment:
We should probably still round the balance when `integer_balance` is `False` up to something like 4 digits.
It can happen in practice that the balance is something like 1.000000000001 because of floating-point arithmetic.